### PR TITLE
[SPARK-23529][K8s] Support mounting hostPath volumes for executors

### PR DESCRIPTION
**Pull request by madanadit**
## What changes were proposed in this pull request?

This PR introduces a new config `spark.kubernetes.executor.volumes` taking a values of the form  `hostPath:containerPath[:ro|rw][,...]`; where  `hostPath` is the path for the executor pod volume, `containerPath` is the mount path and `ro` is read-only mode.

The use case is to enable short-circuit writes to distributed storage on k8s. The Alluxio File System uses domain sockets to enable short-circuit writes from the client to worker memory when co-located on the same host machine. A directory, lets say /tmp/domain on the host, is mounted on the Alluxio worker container as well as the Alluxio client ( = Spark executor) container. The worker creates a domain socket /tmp/domain/d and if the client container mounts the same directory, it can write directly to the Alluxio worker w/o passing through network stack. The end result is faster data access when data is local.

## How was this patch tested?

Manual testing on a k8s v1.8 cluster. Unit tests added to `ExecutorPodFactorySuite`.

### CODE REVIEW

**Comment by madanadit** [comment]
@felixcheung, @foxish can you take a look?

**Comment by foxish** [comment]
Jenkins, ok to test

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 1]
```diff
@@ -117,6 +117,13 @@ private[spark] object Config extends Logging {
       .stringConf
       .createWithDefault("spark")
 
+  val KUBERNETES_EXECUTOR_VOLUMES =
```
Comment (foxish) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala]: If we are adding this option, it should be a more general API - allowing PVs, hostpath volumes and emptyDir volumes to be mounted. In the near future, hostpath volumes will be superseded by local PVs - https://github.com/kubernetes/kubernetes/issues/7562

**Review by foxish — CHANGES_REQUESTED** [review]
Thanks for the PR @madanadit. One initial comment on the approach. I'd also wait for https://github.com/apache/spark/pull/20910 to merge and rebase after that; since that's a pretty large refactor. Also cc/ @mccheah, @kimoonkim this may be interesting to you.

**Comment by liyinan926** [comment]
I think we should have a more general solution so it works for other types of volumes. Particularly, we should have a design discussion on the format of values of such a property so it works for any type of volumes. I suggest we hold on on this PR until we setup a general solution. @foxish @felixcheung .

**Comment by foxish** [comment]
Yes, that is what I was indicating as well. We should have a design for all
different volume types. Extending this PR as the design evolves is a good
way to go about that.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 1 above)
Comment (madanadit) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala]: Yes. I'll add the driver option after I rebase on [#20910](https://github.com/apache/spark/pull/20910) then.

**Comment by liyinan926** [comment]
@madanadit personally I like the first proposal better. The only change I'm proposing is split `spark.kubernetes.executor.volumes.[type].[name].mount` into two keys: `spark.kubernetes.executor.volumes.[type].[name].mount.path` and `spark.kubernetes.executor.volumes.[type].[name].mount.readOnly`, which is `false` by default.

**Comment by foxish** [comment]
@madanadit @liyinan926, can we do a short doc with all the types of volumes and config options and run that through the rest of the community?

**Pull request closed without merging**